README.md
# nowa-build

[![NPM version](https://img.shields.io/npm/v/nowa-build.svg?style=flat)](https://npmjs.org/package/nowa-build)

nowa webpack builder

---

## Feature

- nowa webpack builder

## Install

```bash
$ npm i nowa -g
$ nowa install build
```

## Usage

```bash
$ nowa build --help
```


CHANGELOG.md
# Changelog
## 1.23.0 (2018-10-11)

* [improve] deps be same with nowa-server
* [improve] remove html comments

## 1.22.0 (2018-07-11)

* [feat] --vars & --buildvars support json format

> https://github.com/nowa-webpack/nowa-build/pull/29

## 1.21.0 (2018-04-20)

* [feat] --keepcss flag
> each build creates its own css

## 1.20.1 (2018-04-19)

* [fix] Split css on multi-pages mode.

## 1.20.0 (2017-11-27)

* [feature] Remove include restriction for several loader.

> https://github.com/nowa-webpack/nowa-build/pull/26

## 1.19.1 (2017-11-10)

* [fix] Disable renaming of at-rules such as @keyframes.

> https://github.com/nowa-webpack/nowa-build/issues/25

## 1.19.0 (2017-09-05)

* [feature] Update version of ts-loader.

## 1.18.1 (2017-08-18)

* [fix] CSS autoprefixer support IE9.

> https://github.com/nowa-webpack/nowa-build/issues/23

## 1.18.0 (2017-07-27)

* [feature] Support ES6 uglify.

> https://github.com/nowa-webpack/nowa-build/issues/22

## 1.17.1 (2017-07-06)

* [fix] Fix image path error.

> https://github.com/nowa-webpack/nowa-build/issues/21

## 1.17.0 (2017-06-21)

* [feature] Support IE<=10 `static extends` on `loose` mode.

## 1.16.2 (2017-05-19)

* [improve] Move analyse report file to project dir.

## 1.16.1 (2017-05-18)

* [improve] `--analyse [port]` if port is not specified, open local html.

## 1.16.0 (2017-05-18)

* [feature] Inject `webpack` to user defined config.
* [feature] Add analyse tool, `--analyse`.

## 1.15.2 (2017-05-11)

* [fix] Revert to 1.15.0.

## 1.15.1 (2017-05-10)

* [fix] .ts files support jsx.

## 1.15.0 (2017-03-30)

* [feature] Add `--publicPath` flag.

> https://github.com/nowa-webpack/nowa-build/pull/16

## 1.14.0 (2017-03-16)

* [feature] Add transform-decorators plugin.

> https://github.com/nowa-webpack/nowa/issues/38

## 1.13.0 (2017-03-12)

* [feature] Support page filter to add specified pages to entry.

> https://github.com/nowa-webpack/nowa/issues/35

## 1.12.0 (2017-03-01)

* [improve] Show colored output of build.

> https://github.com/nowa-webpack/nowa-build/issues/15 

## 1.11.0 (2017-02-27)

* [feature] Kill compiler process when webpack fails.

> https://github.com/nowa-webpack/nowa-build/issues/14

## 1.10.0 (2017-01-19)

* [feature] Add TypeScript support.

## 1.9.1 (2017-01-17)

* [fix] Fix loader found error issue.

> https://github.com/nowa-webpack/nowa-server/issues/11

## 1.9.0 (2016-12-26)

* [fix] Fix root dependency mismatch bug.

> https://github.com/nowa-webpack/nowa/issues/30

## 1.8.2 (2016-12-22)

* [fix] Update stylus-loader version for supporting Nodejs 6+.

## 1.8.1 (2016-12-13)

* [fix] Fix alias bug.

## 1.8.0 (2016-12-11)

* [feature] Add alias option for module path alias defination.

> https://github.com/nowa-webpack/nowa-server/issues/10

## 1.7.0 (2016-12-11)

* [feature] Change default image path to `./`.

> https://github.com/nowa-webpack/nowa/issues/27

## 1.6.0 (2016-11-22)

* [feature] Add jsx extension support.

> https://github.com/nowa-webpack/nowa-build/pull/11

## 1.5.0 (2016-11-18)

* [feature] Add mangle option.

## 1.4.0 (2016-10-24)

* [feature] Support font files(woff,woff2,ttf,otf).

## 1.3.2 (2016-09-27)

* [fix] babel-runtime do not support old IEs, so use es3ify-loader to transform babel-runtime.

> https://github.com/nowa-webpack/nowa/issues/22

## 1.3.1 (2016-09-24)

* [feature] Update version of babel plugins.

## 1.3.0 (2016-09-23)

* [feature] Use es2015 internel loose mode.
* [feature] Use babel-runtime to support polyfill.

> https://github.com/nowa-webpack/nowa-build/issues/10

> https://github.com/nowa-webpack/nowa/issues/20

## 1.2.1 (2016-08-24)

* [fix] Fix undefined extension.

## 1.2.0 (2016-08-23)

* [feature] Support multi-compilers via option.
* [feature] Support minify extension.
* [feature] Add include option to support adding more include path for loader.

> https://github.com/nowa-webpack/nowa-build/issues/7

> https://github.com/nowa-webpack/nowa-build/issues/8

> https://github.com/nowa-webpack/nowa-build/issues/9

## 1.1.0 (2016-07-27)

* [feature] Ignore babelrc.
* [improve] Add cache directory for babel to speed up.

> https://github.com/nowa-webpack/nowa-build/issues/5

## 1.0.5 (2016-07-15)

* [fix] Disable zIndex auto-adjustment.

> https://github.com/nowa-webpack/nowa-build/issues/4

## 1.0.4 (2016-07-13)

* [improve] Autoprefixer support low version of Android.

> https://github.com/nowa-webpack/nowa-build/issues/3

## 1.0.3 (2016-07-12)

* [fix] `--loose` argument do not take effect.

> https://github.com/nowa-webpack/nowa-build/issues/2

## 1.0.2 (2016-07-11)

* [fix] Copy task do not keep directory structor.

> https://github.com/nowa-webpack/nowa-build/issues/1

## 1.0.1 (2016-06-30)

* [fix] Can not find nowa plugin when nowa was locally installed.

> https://github.com/nowa-webpack/nowa/issues/6

## 1.0.0 (2016-06-26)

* [improve] Remove peerdependencies of nowa.



src/util.js
/*
* @Author: gbk
* @Date:   2016-05-02 17:15:36
* @Last Modified by:   gbk
* @Last Modified time: 2017-05-18 00:30:12
*/

'use strict';

var fs = require('fs');
var os = require('os');
var path = require('path');
var cp = require('child_process');

var glob = require('glob');
var mkdirp = require('mkdirp');
var chalk = require('chalk');
var webpack = require('webpack');
var minify = require('html-minifier').minify;

var util = {

  // get absolute path to cwd
  cwdPath: function() {
    var argvs = Array.prototype.slice.call(arguments);
    argvs.unshift(process.cwd());
    return path.join.apply(path, argvs);
  },

  // get absolute path to __dirname
  relPath: function(p) {
    var argvs = Array.prototype.slice.call(arguments);
    argvs.unshift(__dirname);
    return path.join.apply(path, argvs);
  },

  // make a webpack entry
  makeEntry: function() {
    return './' + path.join.apply(path, arguments);
  },

  // make all valid pages as webpack entries
  makePageEntries: function(src, entries, pagesFilter) {
    var pages = fs.readdirSync(path.join(src, 'pages'));
    if (typeof pagesFilter === 'string') {
      pagesFilter = pagesFilter.split(',');
      pages = pages.filter(function(page) {
        return pagesFilter.indexOf(page) !== -1;
      });
    }
    pages.forEach(function(page) {
      try {
        var entry = path.join(src, 'pages', page, 'index.js');
        if (fs.statSync(entry).isFile()) {
          entries[page] = util.makeEntry(src, 'pages', page, 'index.js');
          return;
        }
      } catch (e) {
      }
      try {
        var entry = path.join(src, 'pages', page, 'index.jsx');
        if (fs.statSync(entry).isFile()) {
          entries[page] = util.makeEntry(src, 'pages', page, 'index.jsx');
        }
      } catch (e) {
      }
    });
    return entries;
  },

  // merge vars to buildvars
  parseBuildVars: function(vars, buildvars) {
    var key, newVars = {};
    for (key in vars) {
      newVars[key] = [ vars[key] ];
    }
    for (key in buildvars) {
      newVars[key] = buildvars[key];
    }
    return newVars;
  },

  // parse vars for DefinePlugin
  parseVars: function(vars) {
    var newVars = {};
    for (var key in vars) {
      newVars[key] = JSON.stringify(vars[key]);
    }
    return newVars;
  },

  // make filename suffix by vars
  suffixByVars: function (vars, buildvars) {
    if (vars) {
      var suffix = '';
      for (var key in vars) {
        var value = vars[key];

        // filename suffix will not contain `/`
        if (value !== undefined && buildvars[key] && buildvars[key].length > 1) {
          suffix += '-' + value.toString().replace(/\//, '');
        }
      }
      return suffix;
    } else {
      return '';
    }
  },

  // make babel plugin/preset absolute path
  babel: function(type, name) {
    if (Array.isArray(name)) {
      return name.map(function(n) {
        return util.babel(type, n);
      });
    } else {
      if (typeof name === 'object') {
        return [
          require.resolve([
            'babel',
            type,
            name.name
          ].join('-')),
          name.options
        ];
      } else {
        return require.resolve([
          'babel',
          type,
          name
        ].join('-'));
      }
    }
  },

  // copy files to dir
  copy: function(srcPattern, targetDir) {

    if (Array.isArray(srcPattern)) {

      // multi-glob-patterns
      srcPattern.forEach(function(src) {
        util.copy(src, targetDir);
      });

    } else {

      // single-glob-pattern
      var baseDir = srcPattern.split('**')[0];
      glob.sync(srcPattern, {
        nodir: true
      }).forEach(function(file) {

        // read from source and write to dist
        console.log('Copy file: ' + file);
        var target = path.join(targetDir, path.relative(baseDir, file));
        mkdirp.sync(path.dirname(target));
        var content = fs.readFileSync(file);
        if (/\.html?$/.test(file)) {
          content = minify(content.toString('utf-8'), {
            removeComments: true,
            collapseWhitespace: true,
            conservativeCollapse: true,
            preserveLineBreaks: true,
          });
        }
        fs.writeFileSync(target, content);
      });
    }
  },

  // compiler pre-process
  preProcess: function(config) {
    var newConfig;
    try {
      var webpackCfg = require(util.cwdPath('webpack.config.js'));
      newConfig = typeof webpackCfg === 'function' ? webpackCfg(config, webpack) : webpackCfg;
    } catch (e) {
      if (!/Cannot find module.+webpack\.config\.js/.test(e.toString())) {
        console.error(chalk.red('Error in "webpack.config.js"\n' + e.stack));
        process.exit(1);
      }
    }
    return newConfig || config;
  },

  // webpack build failed
  buildFail: function(msg) {
    console.error(chalk.red('\n' + chalk.bold('Webpack Build Failed.') + '\n' + msg));
    process.exit(1);
  },

  // log after finish
  finishLog: function(startStamp) {
    console.log(chalk.green.bold('Finished in ' + ((Date.now() - startStamp) / 1000).toFixed(2) + 's'));
  },
};

module.exports = util;


src/loader.js
/*
* @Author: gbk
* @Date:   2016-05-02 22:07:46
* @Last Modified by:   gbk
* @Last Modified time: 2017-07-06 19:36:37
*/

'use strict';

var os = require('os');
var path = require('path');

var ExtractTextPlugin = require('extract-text-webpack-plugin');

var util = require('./util');

module.exports = function(options, firstRun) {
  var srcPath = util.cwdPath(options.src);
  if (options.includes) {
    srcPath = [
      srcPath
    ].concat(options.includes.map(function(include) {
      return util.cwdPath(include);
    }));
  }
  var exportcss = options.exportcss !== false;
  var presets = util.babel('preset', [
    {
      name: 'es2015',
      options: {
        loose: !!options.loose
      }
    },
    'stage-0',
    'react'
  ]);
  var cacheDirectory = path.join(os.tmpdir(), options.loose ? 'babel-loose' : 'babel-strict');
  var makeLoader = function(remainLoader) {
    var loaders = 'css-loader';
    if (remainLoader) {
      loaders += '!' + remainLoader + '-loader';
    }
    if (exportcss) {
      if (firstRun) {
        return ExtractTextPlugin.extract('style-loader', loaders);
      } else {
        return 'export-css-loader?remove=true&write=false!' + loaders;
      }
    } else {
      return 'style-loader!' + loaders;
    }
  }
  var plugins = [
    'add-module-exports',
    'transform-decorators-legacy',
    'transform-es3-member-expression-literals',
    'transform-es3-property-literals',
    {
      name: 'transform-runtime',
      options: {
        polyfill: !!options.polyfill,
        helpers: false,
        regenerator: true
      }
    }
  ];
  if (options.loose) {
    plugins.push('transform-proto-to-assign');
  }
  return [{
    test: /\.jsx?$/,
    loader: 'babel-loader',
    include: srcPath,
    query: {
      plugins: util.babel('plugin', plugins),
      presets: presets,
      cacheDirectory: cacheDirectory,
      babelrc: false
    }
  }, {
    test: /\.js$/,
    loader: 'es3ify-loader',
    include: function(path) {
      return ~path.indexOf('babel-runtime');
    }
  }, {
    test: /\.css$/,
    loader: makeLoader(),
  }, {
    test: /\.less$/,
    loader: makeLoader('less'),
  }, {
    test: /\.styl$/,
    loader: makeLoader('stylus'),
  }, {
    test: /\.svg$/,
    loader: 'babel-loader',
    query: {
      presets: presets,
      cacheDirectory: cacheDirectory,
      babelrc: false
    }
  }, {
    test: /\.svg$/,
    loader: 'svg2react-loader',
  }, {
    test: /\.json$/,
    loader: 'json-loader',
  }, {
    test: /\.(png|jpe?g|gif|woff|woff2|ttf|otf)$/,
    loader: 'url-loader?limit=10240&publicPath=' + (options.publicPath || './'),
  }, {
    test: /\.tsx?$/,
    loader: 'ts-loader',
    include: srcPath,
    query: {
      transpileOnly: true
    }
  }];
};


src/minify.js
/*
* @Author: gbk
* @Date:   2016-05-10 23:45:10
* @Last Modified by:   gbk
* @Last Modified time: 2017-11-10 16:15:21
*/

'use strict';

var fs = require('fs');

var Balancer = require('load-balancer');
var UglifyJs = require('uglify-es');
var postcss = require('postcss');
var cssnano = require('cssnano');
var processer = new postcss([
  cssnano({
    postcssDiscardUnused: {
      disable: true
    },
    postcssZindex: {
      disable: true
    },
    reduceIdents: false,
    autoprefixer: {
      add: true,
      browsers: [
        '> 1%',
        'IE >= 9',
        'iOS >= 7',
        'Android >= 2.3',
        'FireFoxAndroid >= 46'
      ]
    }
  })
]);

// minify worker
new Balancer.Worker().receive(function(master, context, file, callback) {

  // minify js file
  if (/\.js$/.test(file)) {
    console.log('Minify file: ' + file);
    var result = UglifyJs.minify(fs.readFileSync(file, 'utf-8'), {
      mangle: context.mangle,
      compress: {
        warnings: false,
        drop_console: !context.keepconsole
      },
      output: {
        comments: false,
      }
    });
    fs.writeFileSync(file.replace(/\.js$/, context.minifyExtension + '.js'), result.code);
    callback();

  // minify css file
  } else if (/\.css$/.test(file)) {
    console.log('Minify file: ' + file);
    processer.process(fs.readFileSync(file, 'utf-8'), {
      from: file,
      to: file
    }).then(function(result) {
      fs.writeFileSync(file.replace(/\.css$/, context.minifyExtension + '.css'), result.css);
      callback();
    });

  // in case of files not support
  } else {
    callback();
  }
});


src/index.js
/*
* @Author: gbk <ck0123456@gmail.com>
* @Date:   2016-04-21 17:34:00
* @Last Modified by:   gbk
* @Last Modified time: 2017-05-19 13:22:05
*/

'use strict';

var mkdirp = require('mkdirp');
var webpack = require('webpack');
var ExtractTextPlugin = require('extract-text-webpack-plugin');
var Progress = require('progress');
var Balancer = require('load-balancer');
var BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;

var util = require('./util');
var loader = require('./loader');
var pkg = require('../package.json');

var parseJSONString = function(stringOrAny) {
  if (typeof stringOrAny !== 'string') {
    return stringOrAny;
  } else {
    try {
      return JSON.parse(stringOrAny);
    } catch (e) {
      console.error(e);
      return stringOrAny;
    }
  }
};

// plugin defination
module.exports = {

  description: pkg.description,

  options: [
    [ '-s, --src <dir>', 'source directory, default to `src`', 'src' ],
    [ '-d, --dist <dir>', 'build directory, default to `dist`', 'dist' ],
    [ '-e  --entry <file>', 'app entry, default to `app/app.js`', 'app/app.js' ],
    [ '    --pages [pages]', 'add multi-page entries' ],
    [ '    --buildvars <json>', 'build varibles' ],
    [ '    --vars <json>', 'runtime varibles' ],
    [ '    --externals', 'webpack external varibles' ],
    [ '-o, --loose', 'use babel es2015 loose mode to transform codes' ],
    [ '-c, --keepconsole', 'keep `console.log`' ],
    [ '    --keepcss', 'keep non-extracted css in all builds'],
    [ '    --skipminify', 'skip minify js and css' ],
    [ '-p, --progress', 'show progress' ],
    [ '    --exportcss', 'export css files' ],
    [ '    --multiCompilers', 'generate multi-compilers' ],
    [ '    --minifyExtension <extension>', 'minify file extension' ],
    [ '    --includes', 'loader should include paths' ],
    [ '    --polyfill', 'use core-js to do polyfills' ],
    [ '    --mangle', 'mangle varibles when minify' ],
    [ '    --alias', 'path alias' ],
    [ '    --publicPath <path>', 'set publicPath in webpack'],
    [ '    --analyse [port]', 'analyse bundle size'],
  ],

  action: function(options) {

    // options
    var src = options.src;
    var dist = options.dist;
    var entry = options.entry;
    var pages = options.pages;
    var vars = parseJSONString(options.vars) || {};
    var buildvars = util.parseBuildVars(vars, parseJSONString(options.buildvars) || {});
    var externals = options.externals || {
      'react': 'window.React',
      'react-dom': 'window.ReactDOM || window.React'
    };
    var loose = options.loose;
    var keepconsole = options.keepconsole;
    var keepcss = options.keepcss;
    var skipminify = options.skipminify;
    var showProgress = options.progress;
    var exportcss = options.exportcss !== false;
    var multiCompilers = !!options.multiCompilers;
    var minifyExtension = options.minifyExtension || '';
    var includes = options.includes;
    var polyfill = !!options.polyfill;
    var mangle = !!options.mangle;
    var alias = (function(aliasMap) {
      for (var key in aliasMap) {
        aliasMap[key] = util.cwdPath(src, aliasMap[key]);
      }
      return aliasMap;
    })(typeof options.alias === 'object' ? options.alias : {
      i18n: 'i18n'
    });
    var publicPath = options.publicPath || '/';
    var analyse = options.analyse;

    // start time stamp
    var startStamp = Date.now();

    // copy task
    mkdirp.sync(util.cwdPath(dist));
    util.copy([
      util.cwdPath(src, 'lib', '**', '*.*'),
      util.cwdPath('html', '**', '*.*')
    ], util.cwdPath(dist));

    // entries
    var entries = {
      app: util.makeEntry(src, entry)
    };
    if (pages) {
      entries = util.makePageEntries(src, entries, pages);
    }

    // resolve
    var resolveRoot = [
      util.relPath('..', 'node_modules')
    ];
    if (process.cwd() !== util.relPath('..', '..', '..')) {
      resolveRoot.push(util.relPath('..', '..'));
    }
    var resolve = {
      root: resolveRoot,
      alias: alias,
      extensions: ['', '.js', '.jsx']
    };
    var resolveLoader = {
      root: resolveRoot
    };

    // plugins
    var plugins = [
      new webpack.optimize.OccurenceOrderPlugin()
    ];
    if (exportcss) {
      plugins.push(new ExtractTextPlugin('[name].css')); // allChunks 的会带来css的顺序问题
    }
    if (showProgress) {
      var bar = new Progress('webpack compile [:bar]', {
        total: 1,
        width: 20
      });
      var oldPercentage = 0;
      plugins.push(new webpack.ProgressPlugin(function handler(percentage) {
        bar.tick(percentage - oldPercentage);
        oldPercentage = percentage;
      }));
    }
    if (analyse) {
      var analysePort = parseInt(analyse);
      plugins.push(new BundleAnalyzerPlugin({
        analyzerMode: analysePort > 0 ? 'server' : 'static',
        analyzerPort: analysePort || 8888,
        reportFilename: util.cwdPath('webpack-bundle-analyzer-report.html')
      }));
    }

    // split buildvars into a table
    var keysArr = [];
    var valuesArr = [];
    for (var key in buildvars) {
      keysArr.push(key);
      if (Array.isArray(buildvars[key]) && buildvars[key].length) {
        valuesArr.push(buildvars[key]);
      } else {
        console.error(key + ' in buildvars MUST have at least one value!');
        process.exit(0);
      }
    }

    // generate all vars combination
    var combinations = [];
    if (keysArr.length) {
      var findRoute = function(combination) {
        combination = combination || [];
        var values = valuesArr[combination.length];
        if (values) {
          values.forEach(function(value) {
            findRoute(combination.concat([ value ]));
          });
        } else {
          var vars = {};
          keysArr.forEach(function(key, index) {
            vars[key] = combination[index];
          });
          combinations.push(vars);
        }
      }
      findRoute();
    }

    // minify task
    var minify = skipminify ? function() {
      util.finishLog(startStamp);
    } : function(assets) {
      new Balancer.Master().send(util.relPath('minify.js'), {
        keepconsole: keepconsole,
        minifyExtension: minifyExtension,
        mangle: mangle
      }, assets.map(function(a) {
        return util.cwdPath(dist, a.name);
      }), function() {
        util.finishLog(startStamp);
      });
    };

    // run compiler
    if (combinations.length > 1 || multiCompilers) { // multi-compilers

      var compilers = combinations.length > 1 ? combinations.map(function(vars, index) {
        return util.preProcess({
          entry: entries,
          output: {
            path: util.cwdPath(dist),
            filename: '[name]' + util.suffixByVars(vars, buildvars) + '.js',
            publicPath: publicPath
          },
          plugins: plugins.concat([
            new webpack.DefinePlugin(util.parseVars(vars))
          ]),
          resolve: resolve,
          resolveLoader: resolveLoader,
          externals: externals,
          cache: true,
          module: {
            loaders: loader(options,  keepcss || index === 0)
          }
        });
      }) : util.preProcess({
          entry: entries,
          output: {
            path: util.cwdPath(dist),
            filename: '[name].js',
            publicPath: publicPath
          },
          plugins: plugins,
          resolve: resolve,
          resolveLoader: resolveLoader,
          externals: externals,
          cache: true,
          module: {
            loaders: loader(options, true),
          }
        });

      webpack(compilers, function(err, stats) {

        // print wepack compile result
        if (err) {
          return util.buildFail(err.toString());
        }
        console.log('=============');
        var assets = [];
        stats.stats.forEach(function(stat) {
          if (stat.hasErrors()) {
            return util.buildFail(stat.toJson().errors[0].split('\n').slice(0, 2).join('\n'));
          }
          console.log(stat.toString({
            version: false,
            hash: false,
            chunks: false,
            children: false,
            colors: true
          }) + '\n=============');

          // concat assets
          Array.prototype.push.apply(assets, stat.toJson({
            hash: false,
            chunks: false,
            children: false,
            modules: false
          }).assets);
        });

        // minify task
        minify(assets);
      });

    } else { // single-compiler

      // vars defined
      if (combinations.length) {
        plugins.push(new webpack.DefinePlugin(util.parseVars(combinations[0])));
      }

      webpack(util.preProcess({
        entry: entries,
        output: {
          path: util.cwdPath(dist),
          filename: '[name].js',
          publicPath: publicPath
        },
        plugins: plugins,
        resolve: resolve,
        resolveLoader: resolveLoader,
        externals: externals,
        cache: true,
        module: {
          loaders: loader(options, true)
        }
      }), function(err, stats) {

        // print wepack compile result
        if (err) {
          return util.buildFail(err.toString());
        }
        if (stats.hasErrors()) {
          return util.buildFail(stats.toJson().errors[0].split('\n').slice(0, 2).join('\n'));
        }
        console.log('\n' + stats.toString({
          hash: false,
          chunks: false,
          children: false,
          colors: true
        }));

        // minify task
        minify(stats.toJson({
          hash: false,
          chunks: false,
          children: false,
          modules: false
        }).assets);
      });
    }

  }
};

